File: race_car.cpp
/*************************************************************************
> File Name:  race_car.h
> Author: shenming
> Created Time: Sat Jan 25 08:54:08 2020
************************************************************************/

#include "common_lc.h"

/*
 * state {speed remain} -> {speed*2 remain+speed}
 *                      -> {1 or -1 remain}
 */
class Solution {
public:
    int largest_move;
    map<pair<int, int>, int> mem;

    int get_minimum_steps(pair<int, int> state) {
        if ((state.first > 0 && state.second <0 ) ||
            (state.first <0 && state.second > 0)) {
            int step = log(abs(state.second) + 1)/log(2);
            return step;
        } else {
            int step = log(abs(state.second - 2* state.first))/log(2);
            return step;
        }
    }
    void trace(pair<int, int> state, int step) {
        // printf("[%d %d] %d\n", state.first, state.second, step);
        if (state.second == 0) {
            if (step < largest_move) {
                largest_move = step;
            }
            return;
        }
        //if (mem.count(state)) {
        //    if (mem[state] == INT_MAX) {
        //        return;
        //    } else {
        //        if (step + mem[state] < largest_move) {
        //            largest_move = step + mem[state];
        //        }
        //        return;
        //    }
        //}
        // A
        pair<int, int> a_state(state.first*2, state.second-state.first);
        // printf("a_state: %d, %d\n", a_state.first, a_state.second);
        if (step + 1 + get_minimum_steps(a_state) > largest_move) {
            // do nothing
        } else {
            trace(a_state, step + 1);
        }
        pair<int, int> r_state(state.first > 0 ? -1: 1, state.second);
        if (step + 1+ get_minimum_steps(r_state) > largest_move) {
            // do nothing
        } else {
            trace(r_state, step + 1);
        }
    }

    int racecar(int target) {
        pair<int, int> start(1, target);
        largest_move = 3*target;
        trace(start, 0);
        return largest_move;
    }
};

int main()
{
    Solution s;
    int ret = s.racecar(4);
    printf("%d\n", ret);
    return 0;
}

File: racecar.py
# -*- coding: utf-8 -*-

#####################################################################
# File Name:  racecar.py
# Author: shenming
# Created Time: Sun Jan 26 10:29:12 2020
#####################################################################

import os
import sys

class Solution:
    # why BFS
    # BFS could find minimum value
    def racecar(self, target:int) ->int:
        def bfs():
            from collections import deque
            Q = deque([(0,1,0)])
            seen = {(0,1)}
            if target==0:return 0
            while True:
                p, s, d = Q.popleft()
                for action in [True, False]:
                    if action:
                        this_p = p + s
                        s *= 2
                        if this_p==target:return d+1
                        if (this_p, s) not in seen:
                            seen.add((this_p, s))
                            Q.append((this_p, s, d+1))
                    else:
                        if s>0:s=-1
                        else s=1
                        if (p, s) not in seen:
                            seen.add((p, s))
                            Q.append((p, s, d+1))


if __name__ == "__main__":
    pass
